kernel/kernel-api/src/main/java/io/delta/kernel/internal/ActionCommitLog.java

```diff
@@ -0,0 +1,340 @@
+  static final StructField COMMIT_VERSION_STRUCT_FIELD =
+      new StructField(COMMIT_VERSION_COL_NAME, COMMIT_VERSION_DATA_TYPE, false /* nullable */);
+
+  /** Column name storing the commit timestamp for a given file action */
+  static final String COMMIT_TIMESTAMP_COL_NAME = "timestamp";
+
+  static final DataType COMMIT_TIMESTAMP_DATA_TYPE = LongType.LONG;
+  static final StructField COMMIT_TIMESTAMP_STRUCT_FIELD =
+      new StructField(COMMIT_TIMESTAMP_COL_NAME, COMMIT_TIMESTAMP_DATA_TYPE, false /* nullable */);
+
+  /**
+   * Given a list of delta versions, verifies that they are (1) continuous (2) versions starts with
+   * expectedStartVersion and (3) end with expectedEndVersion. Throws an exception if any of these
+   * are not true.
+   */
+  public static void verifyDeltaVersions(
+      List<FileStatus> commitFiles,
+      Long expectedStartVersion,
+      Long expectedEndVersion,
+      Path tablePath) {
+
+    List<Long> commitVersions =
+        commitFiles.stream()
+            .map(fs -> FileNames.deltaVersion(new Path(fs.getPath())))
+            .collect(Collectors.toList());
+
+    if (!commitVersions.isEmpty()) {
+      List<Long> contVersions =
+          LongStream.rangeClosed(
+                  commitVersions.get(0), commitVersions.get(commitVersions.size() - 1))
+              .boxed()
+              .collect(Collectors.toList());
+      if (!contVersions.equals(commitVersions)) {
+        throw new InvalidTableException(
+            tablePath.toString(),
+            String.format(
+                "Missing delta files: versions are not continuous: (%s)", commitVersions));
+      }
+    }
+
+    if (commitVersions.isEmpty() || !Objects.equals(commitVersions.get(0), expectedStartVersion)) {
+      throw startVersionNotFound(
+          tablePath.toString(),
+          expectedStartVersion,
+          commitVersions.isEmpty() ? Optional.empty() : Optional.of(commitVersions.get(0)));
+    }
+
+    if (commitVersions.isEmpty()
+        || !Objects.equals(commitVersions.get(commitVersions.size() - 1), expectedEndVersion)) {
+      throw endVersionNotFound(
+          tablePath.toString(),
+          expectedEndVersion,
+          commitVersions.isEmpty()
+              ? Optional.empty()
+              : Optional.of(commitVersions.get(commitVersions.size() - 1)));
+    }
+  }
+
+  /**
+   * Gets an iterator of files in the _delta_log directory starting with the startVersion.
+   *
+   * @throws TableNotFoundException if the directory does not exist
+   */
+  private static CloseableIterator<FileStatus> listLogDir(
+      Engine engine, Path tablePath, long startVersion) {
+    final Path logPath = new Path(tablePath, "_delta_log");
+    try {
+      return wrapEngineExceptionThrowsIO(
+          () ->
+              engine.getFileSystemClient().listFrom(FileNames.listingPrefix(logPath, startVersion)),
+          "Listing from %s",
+          FileNames.listingPrefix(logPath, startVersion));
```

Comment:
It looks like we need to construct a snapshot in order to list unbackfilled commits from the coordinated commit enabled table.

Reply:
not just constructing a latest snapshot, but keeping it updated with the latest changes as the `Table` instance is reused across the streaming batches. May be for Delta Sharing this isn't that important.